### PR TITLE
cmd/snap-confine: use snap-discard-ns ns to discard stale namespaces

### PR DESCRIPTION
**Pull request by zyga**
During startup snap-confine may decide to discard a preserved mount
namespace that became stale. This used to involve unmounting one file.
Then we realized we had a bug because the .fstab file was not unlinked.
With the introduction of per-user mount namespaces we need to do a little
bit more. Instead of duplicating this logic or providing library code
I believe the easiest way is to execute snap-discard-ns itself.

The same patch contains a small refactoring for opening internal tools.

Signed-off-by: Zygmunt Krynicki <me@zygoon.pl>

### CODE REVIEW

**Comment by zyga** [comment]
NOTE: this depends on https://github.com/snapcore/snapd/pull/6150 and https://github.com/snapcore/snapd/pull/6145

**Comment by zyga** [comment]
This is failing because .mnt files are now removed for real and the code doesn't cope with that yet. I will address this now.

File: cmd/snap-confine/ns-support.c [hunk 1]
```diff
@@ -307,18 +307,59 @@ enum sc_discard_vote {
 	SC_DISCARD_YES = 2,
 };
 
+static void call_snap_discard_ns(int snap_discard_ns_fd, const char *snap_name)
```
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-confine/ns-support.c]: Hmm, not sure why? Can you explain please?

File: cmd/snap-confine/ns-support.c [hunk 2]
```diff
@@ -654,6 +683,14 @@ static void helper_capture_ns(struct sc_mount_ns *group, pid_t parent)
 	debug("capturing per-snap mount namespace");
 	sc_must_snprintf(src, sizeof src, "/proc/%d/ns/mnt", (int)parent);
 	sc_must_snprintf(dst, sizeof dst, "%s%s", group->name, SC_NS_MNT_FILE);
+	if (access(dst, F_OK) < 0 && errno == ENOENT) {
+		int fd = open(dst, O_CREAT | O_CLOEXEC | O_NOFOLLOW | O_RDONLY,
+			      0600);
+		if (fd < 0) {
+			die("cannot open file: %s", dst);
```
Comment (chipaca) [inline cmd/snap-confine/ns-support.c]: more like 'cannot create', no?

File: cmd/snap-confine/snap-confine.apparmor.in [hunk 3]
```diff
@@ -451,26 +451,26 @@
...
 
-    # Allow executing snap-update-ns when...
+    # Allow executing snap-update-ns and snap-discard-ns when...
 
     # ...snap-confine is, conceptually, re-executing and uses snap-update-ns
     # from the distribution package. This is also the location used when using
     # the core/base snap on all-snap systems. The variants here represent
     # various locations of libexecdir across distributions.
-    /usr/lib{,exec,64}/snapd/snap-update-ns r,
+    /usr/lib{,exec,64}/snapd/snap-{update,discard}-ns rix,
```
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-confine/snap-confine.apparmor.in]: Not idea really. We open and run both via descriptors. One is ran before pivot_root. Perhaps in this case we can actually figure out what is being executed so this is needed.

[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-confine/snap-confine.apparmor.in]: I started investigating this and I cannot reproduce the need for the extra permissions. I may revert that bit.

[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-confine/snap-confine.apparmor.in]: Spoke too soon:
```
2018-11-27 08:48:46 Error executing google:ubuntu-16.04-64:tests/main/stale-base-snap : 
-----
+ pid=11843
++ seq 30
+ test-snapd-sh -c 'touch $SNAP_DATA/stamp && exec sleep 1h'
+ for _ in '$(seq 30)'
+ test -e /var/snap/test-snapd-sh/current/stamp
+ sleep 0.1
+ for _ in '$(seq 30)'
+ test -e /var/snap/test-snapd-sh/current/stamp
+ sleep 0.1
+ for _ in '$(seq 30)'
+ test -e /var/snap/test-snapd-sh/current/stamp
+ break
+ test-snapd-sh -c 'touch /tmp/stamp'
+ snap install --dangerous ./core-customized.snap
core customized installed
+ MATCH customized
+ snap list
+ test-snapd-sh -c 'test -e /tmp/stamp'
+ test-snapd-sh -c 'test ! -e /customized'
+ kill 11843
+ wait -n
/bin/bash: line 87: 11843 Terminated              test-snapd-sh -c 'touch $SNAP_DATA/stamp && exec sleep 1h'
+ true
+ test-snapd-sh -c 'test ! -e /tmp/stamp'
cannot execute snap-discard-ns: Permission denied
snap-discard-ns failed with code 1
-----
```

The denial is:

```
Nov 27 07:48:45 nov270732-548913 audit[12178]: AVC apparmor="DENIED" operation="exec" profile="/snap/core/x1/usr/lib/snapd/snap-confine" name="/snap/core/x1/usr/lib/snapd/snap-discard-ns" pid=12178 comm="snap-confine" requested_mask="x" denied_mask="x" fsuid=0 ouid=0
```

[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-confine/snap-confine.apparmor.in]: Perhaps the key is this: we change to snap-update-ns unique profile when calling that binary. For snap-discard-ns we do not.  The s-u-n profile contains this:
```
  /usr/lib{,exec,64}/snapd/snap-update-ns mr,
  /var/lib/snapd/hostfs/usr/lib{,exec,64}/snapd/snap-update-ns mr,
  /{,var/lib/snapd/}snap/{core,snapd}/*/usr/lib/snapd/snap-update-ns mr,
  /var/lib/snapd/hostfs/{,var/lib/snapd/}snap/core/*/usr/lib/snapd/snap-update-ns mr,
```

File: cmd/snap-confine/ns-support.c [hunk 4]
```diff
@@ -647,6 +676,16 @@ static void helper_capture_ns(struct sc_mount_ns *group, pid_t parent)
 	debug("capturing per-snap mount namespace");
 	sc_must_snprintf(src, sizeof src, "/proc/%d/ns/mnt", (int)parent);
 	sc_must_snprintf(dst, sizeof dst, "%s%s", group->name, SC_NS_MNT_FILE);
+	if (access(dst, F_OK) < 0 && errno == ENOENT) {
+		int fd =
+		    open(dst,
```
Comment (chipaca) [inline cmd/snap-confine/ns-support.c]: out of curiosity, why wasn't this written as
```c
int fd = open(...);
if (fd < 0 && errno != EEXIST) { /* error out */ }
```
?

Comment (zyga) [reply cmd/snap-confine/ns-support.c]: No specific reason really. I guess I can simplify.

File: cmd/snap-confine/ns-support.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-confine/ns-support.c]: I did this refactoring but it turned out to be larger than I hoped. I will propose this separately next.

**Pull request merged**